src/components/Image/Image.tsx

```diff
@@ -7,34 +14,104 @@ export interface SourceSet {
 
 export type CrossOrigin = 'anonymous' | 'use-credentials' | '' | undefined;
 
+export interface State {
+  hasError: boolean;
+  hasLoaded: boolean;
+}
+
 export interface Props extends React.HTMLProps<HTMLImageElement> {
   alt: string;
+  className?: string;
   source: string;
   sourceSet?: SourceSet[];
+  onError?(): void;
 }
 
-export default function Image({
-  sourceSet,
-  source,
-  crossOrigin,
-  ...rest
-}: Props) {
-  const finalSourceSet = sourceSet
-    ? sourceSet
-        .map(({source: subSource, descriptor}) => `${subSource} ${descriptor}`)
-        .join(',')
-    : null;
-
-  return finalSourceSet ? (
-    // eslint-disable-next-line jsx-a11y/alt-text
-    <img
-      src={source}
-      srcSet={finalSourceSet}
-      crossOrigin={crossOrigin as CrossOrigin}
-      {...rest}
-    />
-  ) : (
-    // eslint-disable-next-line jsx-a11y/alt-text
-    <img src={source} {...rest} crossOrigin={crossOrigin as CrossOrigin} />
-  );
+export type CombinedProps = Props & WithAppProviderProps;
+
+export class Image extends React.PureComponent<CombinedProps, State> {
+  state: State = {
+    hasError: false,
+    hasLoaded: false,
+  };
+
+  ComponentDidUpdate({
+    source: previousSource,
+    sourceSet: previousSourceSet = [],
+  }: Props) {
+    const {source, sourceSet = []} = this.props;
+    if (
+      previousSource !== source ||
+      !arraysAreEqual(sourceSet, previousSourceSet)
+    ) {
+      this.setState({hasError: false, hasLoaded: false});
+    }
+  }
+
+  render() {
+    const {
+      sourceSet,
+      source,
+      crossOrigin,
+      onError,
+      className,
+      ...rest
+    }: Props = this.props;
+    const {hasError = false, hasLoaded = false} = this.state;
+
+    if (hasError) {
+      return null;
```

Comment:
I don’t think we should do this or any of the load and error state tracking. I would pass the `onError` handler directly to `img` and leave it at that. It feels like we are trying to do too much and smells weird to me.

Reply:
I don't see how we'd get around this.  The nice this thing is that someone could provide a custom image for images that are not found.